tests/publiccloud/cloud_netconfig.pm

```diff
@@ -12,77 +12,107 @@
+        $query_meta_ipv4_cmd = qq(curl -H "Metadata-Flavor: Google" "http://$pc_meta_api_ip/computeMetadata/v1/instance/network-interfaces/0/ip");
+    } else {
+        die("Unsupported public cloud provider");
     }
 
-    $instance->ssh_assert_script_run(
-        cmd => 'systemctl status cloud-netconfig.timer',
-        fail_message => 'cloud-netconfig.timer appears not to be started.'
-    );
+    my $data = $instance->ssh_script_output($query_meta_ipv4_cmd);
+
+    return $data;
+}
+
+sub run {
+    my ($self, $args) = @_;
+    my $instance = $args->{my_instance};
+    $self->{instance} = $args->{my_instance};
+    my $pers_net_rules = '/etc/udev/rules.d/75-persistent-net-generator.rules';
+
+    $instance->ssh_assert_script_run('systemctl is-enabled cloud-netconfig.service');
+
+    $instance->ssh_assert_script_run('systemctl status cloud-netconfig.timer');
 
     # 75-persistent-net-generator.rules is usually a symlink to /dev/null in SLES 12+
-    $instance->ssh_assert_script_run(
-        cmd => "test -L $pers_net_rules",
-        fail_message => "File \"$pers_net_rules\" should be a symlink to "
-          . "\"/dev/null\" in SLES 12+."
-    );
-
-    my $grep_cloudvm_ipv4_cmd =
-      'ip -4 addr show eth0 | grep -oP "^\s+inet\s(\d+\.){3}\d+" | '
-      . 'sed "s/ *inet //g"';
-    my $local_eth0_ip =
-      $instance->ssh_script_output(cmd => $grep_cloudvm_ipv4_cmd);
+    $instance->ssh_assert_script_run("test -L $pers_net_rules");
+
+    record_info('eth0');
+    my $local_eth0_ip = $instance->ssh_script_output(qq(ip -4 -o a s eth0 | grep -Po "inet \\K[\\d.]+" | head -n1));
     chomp($local_eth0_ip);
+    my $metadata_eth0_ip = $self->query_metadata($instance, ifNum => '0', addrCount => '0');
+    die("Failed to get interface 0 IP from metadata server") unless length($metadata_eth0_ip);
+    assert_equals($metadata_eth0_ip, $local_eth0_ip, 'Locally assigned eth1 IP does not equal the IP retrieved from CSP metadata service.');
 
-    my $query_meta_ipv4_cmd = "";
-    if (is_azure()) {
-        $query_meta_ipv4_cmd =
-          'curl -H Metadata:true '
-          . "\"http://$pc_meta_api_ip/metadata/instance/network/interface/0/"
-          . 'ipv4/ipAddress/0/privateIpAddress?api-version=2023-07-01'
-          . '&format=text"';
-    }
-    elsif (is_ec2()) {
-        my $access_token = $instance->ssh_script_output(cmd => "curl -X PUT http://$pc_meta_api_ip/latest/api/token -H \"X-aws-ec2-metadata-token-ttl-seconds: 60\"");
-        record_info("DEBUG", $access_token);
-        $query_meta_ipv4_cmd =
-          "curl -H \"X-aws-ec2-metadata-token: $access_token\" \"http://$pc_meta_api_ip/latest/meta-data/local-ipv4\"";
-    }
-    elsif (is_gce()) {
-        $query_meta_ipv4_cmd =
-          'curl -H "Metadata-Flavor: Google" '
-          . "\"http://$pc_meta_api_ip/computeMetadata/v1/instance/"
-          . 'network-interfaces/0/ip"';
-    } else {
-        die("Unsupported public cloud provider.");
+    if (is_azure) {
```

Comment:
same here , whenever there is a pattern : 
```
if (is_azure()) {
} elsif(is_ec2){
} elsif(is_gce) {
}
```
it should be split between dedicated provider children . this is why we have this structure in the first place , to avoid such constructions in each test